in_toto/keylib.go

```diff
@@ -2,220 +2,402 @@ package in_toto
+			return err
+		}
+	case *rsa.PrivateKey:
+		// Note: We store the public key as PKCS8 key here, although the private key get's stored as PKCS1 key
+		// This behavior is consistent to the securesystemslib
+		pubKeyBytes, err := x509.MarshalPKIXPublicKey(key.(*rsa.PrivateKey).Public())
+		if err != nil {
+			return err
+		}
+		if err := k.SetKeyComponents(pubKeyBytes, pemBytes, "rsa", scheme, keyIdHashAlgorithms); err != nil {
+			return err
+		}
+	case ed25519.PublicKey:
+		if err := k.SetKeyComponents(key.(ed25519.PublicKey), []byte{}, "ed25519", scheme, keyIdHashAlgorithms); err != nil {
+			return err
+		}
+	case ed25519.PrivateKey:
+		pubKeyBytes := key.(ed25519.PrivateKey).Public()
+		if err := k.SetKeyComponents(pubKeyBytes.(ed25519.PublicKey), key.(ed25519.PrivateKey), "ed25519", scheme, keyIdHashAlgorithms); err != nil {
+			return err
+		}
+	case *ecdsa.PrivateKey:
+		pubKeyBytes, err := x509.MarshalPKIXPublicKey(key.(*ecdsa.PrivateKey).Public())
+		if err != nil {
+			return err
+		}
+		if err := k.SetKeyComponents(pubKeyBytes, pemBytes, "ecdsa", scheme, keyIdHashAlgorithms); err != nil {
+			return err
+		}
+	case *ecdsa.PublicKey:
+		if err := k.SetKeyComponents(pemBytes, []byte{}, "ecdsa", scheme, keyIdHashAlgorithms); err != nil {
+			return err
+		}
+	default:
+		return fmt.Errorf("%w: %T", ErrUnsupportedKeyType, key)
+	}
+	return nil
 }
 
 /*
-GenerateEd25519Signature creates an ed25519 signature using the key and the
-signable buffer provided. It returns an error if the underlying signing library
-fails.
-*/
-func GenerateEd25519Signature(signable []byte, key Key) (Signature, error) {
+GenerateSignature will automatically detect the key type and sign the signable data
+with the provided key. If everything goes right GenerateSignature will return
+a for the key valid signature and err=nil. If something goes wrong it will
+return an not initialized signature and an error. Possible errors are:
 
-	var signature Signature
+	* ErrNoPEMBlock
+	* ErrUnsupportedKeyType
 
-	seed, err := hex.DecodeString(key.KeyVal.Private)
-	if err != nil {
-		return signature, err
+*/
+func GenerateSignature(signable []byte, key Key) (Signature, error) {
+	var signature Signature
+	var signatureBuffer []byte
+	// The following switch block is needed for keeping interoperability
+	// with the securesystemslib and the python implementation
+	// in which we are storing RSA keys in PEM format, but ed25519 keys hex encoded.
+	switch key.KeyType {
+	case "rsa", "ecdsa":
+		// pam.Decode returns the parsed pem block and a rest.
+		// The rest is everything, that could not be parsed as PEM block.
+		// Therefore we can drop this via using the blank identifier "_"
+		data, _ := pem.Decode([]byte(key.KeyVal.Private))
+		if data == nil {
+			return signature, ErrNoPEMBlock
+		}
+		parsedKey, err := ParseKey(data.Bytes)
+		if err != nil {
+			return signature, err
+		}
```

Comment:
It might enhance readability, if you create a small helper function for the 8 lines above, remove the inner switch, and just call that helper (and sha256.Sum256(signable)) twice.  What do you think? 

Btw. you could also use that helper in load key, where you have the same 8 lines.

Reply:
The latest commit should contain a first draft of this. The problem I see here:

1. We don't use iotuil.ReadAll in the other functions, so we can't get this into the helper function. Therefore the lines get reduced on 5, instead of 8.
2. I still see a need for the inner switch, because the inner switch makes sure that we use the right key for the right key operations. Right now it is possible to load a key with keyType "rsa"  with a PKCS8-ECDSA key. In the current state, the inner switch will make sure, that we use the correct operation for the key. If we remove this inner type switch, we could run into a panic, if somebody creates a keyobject with non-matching keytype and key.

I am not sure yet, if we should add such keyType/key checking to our functions in model, or if we should keep it this way and let the functions be as defensive as possible. What I could think of is: We add a check for this in our validateKey function in model.go and call this function in the first line of GenerateSignature and ValidateSignature. This way we would make sure, that we always operate on a valid object. What do you think?

Reply:
Type/Scheme-checking everything done in the latest commit :D